src/tests/unit_asio_stream.cpp

```diff
@@ -0,0 +1,560 @@
+#include "tests.h"
+
+#include <iostream>
+
+#if defined(BOTAN_HAS_TLS) && defined(BOTAN_HAS_BOOST_ASIO)
+
+#include <botan/asio_stream.h>
+#include <botan/tls_callbacks.h>
+
+#include <boost/beast/experimental/test/stream.hpp>
+#include <boost/bind.hpp>
+
+namespace Botan_Tests {
+
+namespace asio   = boost::asio;
+using error_code = boost::system::error_code;
+
+constexpr uint8_t     TEST_DATA[] = "The story so far: In the beginning the Universe was created. "
+                                    "This has made a lot of people very angry and been widely regarded as a bad move.";
+constexpr std::size_t TEST_DATA_SIZE = 142;
+static_assert(sizeof(TEST_DATA) == TEST_DATA_SIZE, "size of TEST_DATA must match TEST_DATA_SIZE");
+
+// use memcmp to check if the data in a is a prefix of the data in b
+bool contains(const void* a, const void* b, const std::size_t size) { return memcmp(a, b, size) == 0; }
+
+/**
+ * Mocked Botan::TLS::Channel. Pretends to perform TLS operations and triggers appropriate callbacks in StreamCore.
+ */
+class MockChannel
```

Comment:
If you have access to beast (or if you can just copy the code) you might want to try this stream as your next layer:
https://github.com/boostorg/beast/blob/develop/include/boost/beast/_experimental/test/stream.hpp#L37

It is quite solid, it can simulate failures (great for testing all branches and achieving high coverage), and you can "stuff" the stream with bytes. The Beast tests make heavy use of it. In this code, you can see how the test stream is stuffed with an HTTP request, and the call to `http::read` is run in a "failure loop" which fails at the `n` I/O operation this is how Beast gets close to 100% coverage:
https://github.com/boostorg/beast/blob/develop/test/beast/http/read.cpp#L277